assume/units/storage.py

```diff
@@ -408,51 +349,53 @@ def set_dispatch_plan(
         end: pd.Timestamp,
         product_type: str,
     ):
-        end_excl = end - self.index.freq
-        self.outputs["energy"].loc[start:end_excl] += dispatch_plan["total_power"]
+        duration = (end - start).total_seconds() / 3600
 
         # TODO checks should be at execute_current_dispatch - see powerplant
         # TODO check if resulting power is < max_power
-        # if self.outputs["energy"][start:end_excl].max() > self.max_power:
-        #     max_pow = self.outputs["energy"][start:end_excl].max()
+        # if self.outputs["energy"][start:end].max() > self.max_power:
+        #     max_pow = self.outputs["energy"][start:end].max()
         #     logger.error(f"{max_pow} greater than {self.max_power} - bidding twice?")
 
-        if self.outputs["energy"][start:end_excl].min() >= self.min_power_discharge:
-            self.market_success_list[-1] += 1
-            self.current_status = 1  # discharging
-            self.current_down_time = 0
-            self.outputs["energy"].loc[start:end_excl] = dispatch_plan["total_power"]
-            self.current_SOC = (
-                self.current_SOC
-                - dispatch_plan["total_power"] / self.efficiency_discharge
+        if self.outputs["energy"][start:end].min() >= self.min_power_discharge:
+            self.set_market_sucess(dispatch_plan, start, end)
+            discharged_energy = (
+                dispatch_plan["total_power"] * duration / self.efficiency_discharge
             )
+            self.current_SOC = self.current_SOC - discharged_energy
 
-        elif self.outputs["energy"][start:end_excl].max() <= -self.min_power_charge:
-            self.market_success_list[-1] += 1
-            self.current_status = 1  # charging
-            self.current_down_time = 0
-            self.outputs["energy"].loc[start:end_excl] = dispatch_plan["total_power"]
-            self.current_SOC = (
-                self.current_SOC - dispatch_plan["total_power"] * self.efficiency_charge
+        elif self.outputs["energy"][start:end].max() <= -self.min_power_charge:
+            self.set_market_sucess(dispatch_plan, start, end)
+            # multiplied by -1 because the power when charging is negative
+            charged_energy = (
+                dispatch_plan["total_power"] * duration * self.efficiency_charge * (-1)
             )
+            self.current_SOC = self.current_SOC + charged_energy
 
-        elif self.outputs["energy"][start:end_excl].min() < self.min_power_discharge:
+        else:
             self.current_status = 0
             self.current_down_time += 1
-            self.outputs["energy"].loc[start:end_excl] = 0
+            self.outputs["energy"].loc[start:end] = 0
```

Comment:
I am quite sure that end_excl was needed in the section above..?

power
10:00 - 40
11:00 - 50
12:00 - 60

now start is 10:00 and end is 11:00 for our bid
if you set  self.outputs["energy"].loc[start:end] = 0 here - you would set the dispatch for the time from 11-12 to 0 too, which is wrong - same for other occurances of this behavior.

The end has to be exclusive. So the previous behavior was more correct?


Additionally, this does not work with multiple markets here - the checks should go into `execute_current_dispatch`

Reply:
I believe the assumption to have the same value for the next time step can still be valid as it helps the bidding strategies. Also, the correct and final value will be set after the respective market clearing. 
So I sold 10MW form 10 to 11, so for next bid calculation I don't want to assume my currrent power as 0, since this is not true, but 10, as this was my last state. So I use this 10 as current power and maybe sell now 15. But I see also some other points where this leads to errors. We should discuss this